README
NAME
    Stepford - A vaguely Rake/Make/Cake-like thing for Perl - create steps
    and schedule them

VERSION
    version 0.000002

SYNOPSIS
        package My::Step::MakeSomething;

        use Moose;

        with 'StepFord::Role::Step::FileGenerator';

        sub run {

            # write some files somehow
        }

        package My::Runner;

        use Stepford::Scheduler;

        my @steps = (
            My::Step::Step1->new(
                name => 'step 1',
                ...
            ),
            My::Step::Step2->new(
                name => 'step 2',
                ...
            ),
            My::Step::MakeSomething->new(
                name         => 'Generate a file',
                dependencies => [ 'step 1', 'step 2' ],
            ),
        );

        my $target_step = $steps[-1];

        # Runs all the steps needed to get to the $final_step.
        Stepford::Scheduler->new(
            steps => \@steps,
        )->run($final_step);

DESCRIPTION
    NOTE: This is some alpha ju^H^Hcode. You have been warned!

    Stepford provides a framework for running a set of steps that are
    dependent on other steps. At a high level, this is a lot like Make,
    Rake, etc. However, the actual implementation is fairly different.
    Currently, there is no DSL, no Stepfile, etc.

    With Stepford, each step is represented by a class you create. That
    class should consume either the StepFord::Role::Step::FileGenerator role
    (if it generates files) or the StepFord::Role::Step step (if it
    doesn't).

    You then instantiate step objects for each step, giving each step a name
    and explicitly specifying its dependencies. Finally, you pass all these
    steps to a Stepford::Scheduler and tell it to run a given step. The
    scheduler runs all the dependent steps (and their dependencies and so
    on).

    Each step can specify a "last_run_time()" method (or get one from the
    StepFord::Role::Step::FileGenerator role). The scheduler uses this to
    skip steps that are up to date.

    See Stepford::Scheduler, Stepford::Role::Step, and
    StepFord::Role::Step::FileGenerator for more details.

FUTURE FEATURES
    There are several very obvious things that should be added to this
    framework:

    *   Logging

        The scheduler and steps should all accept some sort of optional log
        object and tell it what they're doing.

    *   Dry runs

        This requires logging, of course.

    *   Parallel running

        Since the scheduler know what steps depend on what other steps, it
        can also figure out when things can be run in parallel.

VERSIONING POLICY
    This module uses semantic versioning as described by
    <http://semver.org/>. Version numbers can be read as X.YYYZZZ, where X
    is the major number, YYY is the minor number, and ZZZ is the patch
    number.

SUPPORT
    Please report all issues with this code using the GitHub issue tracker
    at <https://github.com/maxmind/Stepford/issues>.

AUTHOR
    Dave Rolsky <drolsky@maxmind.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by MaxMind, Inc..

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
0.003001 2014-12-17

[BUG FIXES]

- Remove use of Path::Class::File->copy_to method in a test, as this requires
  new(ish) version of Path::Class (0.33), but nothing else in this distro
  would require a specific new version (that we know of).


0.003000 2014-12-02

[BACKWARDS INCOMPATIBILITIES]

- Renamed Stepford::Planner to Stepford::Runner. The old class name will still
  work, but you will get a warning every time you call Stepford::Planner->new.

[ENHANCEMENTS]

- Added the Stepford::Role::Step::Unserializable role. This role tells the
  Stepford::Runner not to run the step class in a forked process because it's
  productions cannot be serialized. Instead, this step will be run in the same
  process as the planner.


0.002011 2014-11-13

- The Stepford::Role::Step::FileGenerator::Atomic now creates its
  pre_commit_file() in the same directory as the final file. Previously we
  used a temp dir but renaming the pre_commit_file() to the final file could
  fail if the two files were on different filesystems.


0.002010 2014-11-12

- Fixed a bug with Stepford::Role::Step::FileGenerator::Atomic steps and
  parallel test runs. The step would create a temp directory in the parent
  process which could be deleted by the time the step's run() method was
  called in a child process.


0.002009 2014-10-22

- There is now a new Stepford::Role::Step::FileGenerator::Atomic role. This
  role can be consumed by step classes which want to ensure that single file
  is generated atomically. Implemented by Ran Eilam. GH #11.


0.002008 2014-10-14

- (Some) previous releases were missing a LICENSE file.


0.002007 2014-09-30

- FileGenerator steps can now return undef from the last_run_time() method in
  order to indicate that the step must be re-run. The last_run_time() method
  provided by the Stepford::Role::Step::FileGenerator role does this when any
  of the production files are missing. Fixed by Ran Eilam. GH #9.

- When running a parallel plan, steps that should have been rebuilt because of
  stale files were not being rebuilt properly. Fixed by Ran Eilam. GH #10.


0.002006 2014-08-12

- The logger passed to each step class is now wrapped to prepend a per-step
  class moniker to each log message. This makes following the logs for a
  parallel run much easier.

- The planner object now recognizes inner classes created by any files that it
  loads. This means you can create step classes on the fly and the planner
  will recognize them as being valid.


0.002005 2014-06-30

- More doc fixes for Stepford::Runner. The docs on when we do error checks
  and what type of errors are checked for were out of date.


0.002004 2014-06-30

- Fixed the Stepford::Runner docs. The docs said that the constructor
  accepted a final_steps method, and did not document the parameter passed to
  the run() method properly. Bug report by Olaf Alders. GH #5.


0.002003 2014-06-19

- Added debug level logging that explains how each dependency for each class
  is resolved.


0.002002 2014-06-02

- The Stepford::Runner object now throws an error if it finds a class in any
  given step namespace that is a subclass of Moose::Object but which doesn't
  consume the Stepford::Role::Step role.


0.002001 2014-05-21

- Fixed an incorrect example in the Stepford::Runner SYNOPSIS and improved
  the larger example in the Stepford.pm SYNOPSIS.


0.002000 2014-05-21

* Breaking API changes. Final step(s) are now passed to the $planner->run()
  method rather than the Stepford::Runner constructor.

- Implemented the ability to specify multiple final steps for a run.

- Implemented parallel running.

- All runs now end with a new step, Stepford::FinalStep. This just logs a
  message saying that the run is complete.

- Plans are now optimized so steps run as early as possible in the plan. This
  helps improve parallelism and is just more generally correct. Thanks to Ran
  Eilam for pointing me at a blog post
  (blog.codeaholics.org/parallel-ant/#how-it-works) that described a very
  simple algorithm to do this.


0.001002 2014-04-22

- Require Time::HiRes 1.9726. Some earlier versions broke constructs like
  (stat $file)[9]. This construct is used in the Stepford::Role::FileGenerator
  role.


0.001001 2014-03-27

- (Hopefully) Fix some test failures that look they're caused by hash
  randomization. Reported by Slaven Rezic. GitHub issue #1.


0.001000 2014-03-20

- Totally redesigned the whole thing after discussions with Greg Oschwald here
  at MaxMind. I did say this was alpha!


0.000002 2014-03-03

- The tests now require Time::HiRes 1.9722. It seems like the version shipped
  before Perl 5.16 (1.9721_01) has a weird bug when used like this:

      some_sub( (stat $file)[9] );

  It seems to return the value "9" rather than the mtime.


0.000001 2014-02-25

- First release upon an unsuspecting world.


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.029.
use strict;
use warnings;

use 5.010;

use ExtUtils::MakeMaker;



my %WriteMakefileArgs = (
  "ABSTRACT" => "A vaguely Rake/Make/Cake-like thing for Perl - create steps and let a runner run them",
  "AUTHOR" => "Dave Rolsky <drolsky\@maxmind.com>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Stepford",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.010",
  "NAME" => "Stepford",
  "PREREQ_PM" => {
    "Carp" => 0,
    "File::Temp" => 0,
    "Forest::Tree" => 0,
    "List::AllUtils" => 0,
    "Log::Dispatch" => 0,
    "Log::Dispatch::Null" => 0,
    "Module::Pluggable::Object" => 0,
    "Moose" => 0,
    "Moose::Role" => 0,
    "MooseX::Params::Validate" => 0,
    "MooseX::StrictConstructor" => 0,
    "MooseX::Types" => 0,
    "MooseX::Types::Combine" => 0,
    "MooseX::Types::Common::Numeric" => 0,
    "MooseX::Types::Common::String" => 0,
    "MooseX::Types::Moose" => 0,
    "MooseX::Types::Path::Class" => 0,
    "Parallel::ForkManager" => 0,
    "Path::Class" => 0,
    "Scalar::Util" => 0,
    "Scope::Guard" => 0,
    "Throwable::Error" => 0,
    "Time::HiRes" => "1.9726",
    "namespace::autoclean" => 0,
    "parent" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Copy" => 0,
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "IPC::Signal" => 0,
    "Log::Dispatch::Array" => 0,
    "Test::Differences" => 0,
    "Test::Fatal" => 0,
    "Test::More" => "0.88",
    "Test::Requires" => 0,
    "Test::Warnings" => 0,
    "Try::Tiny" => 0,
    "autodie" => 0,
    "lib" => 0
  },
  "VERSION" => "0.003001",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Copy" => 0,
  "File::Spec" => 0,
  "File::Temp" => 0,
  "Forest::Tree" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "IPC::Signal" => 0,
  "List::AllUtils" => 0,
  "Log::Dispatch" => 0,
  "Log::Dispatch::Array" => 0,
  "Log::Dispatch::Null" => 0,
  "Module::Pluggable::Object" => 0,
  "Moose" => 0,
  "Moose::Role" => 0,
  "MooseX::Params::Validate" => 0,
  "MooseX::StrictConstructor" => 0,
  "MooseX::Types" => 0,
  "MooseX::Types::Combine" => 0,
  "MooseX::Types::Common::Numeric" => 0,
  "MooseX::Types::Common::String" => 0,
  "MooseX::Types::Moose" => 0,
  "MooseX::Types::Path::Class" => 0,
  "Parallel::ForkManager" => 0,
  "Path::Class" => 0,
  "Scalar::Util" => 0,
  "Scope::Guard" => 0,
  "Test::Differences" => 0,
  "Test::Fatal" => 0,
  "Test::More" => "0.88",
  "Test::Requires" => 0,
  "Test::Warnings" => 0,
  "Throwable::Error" => 0,
  "Time::HiRes" => "1.9726",
  "Try::Tiny" => 0,
  "autodie" => 0,
  "lib" => 0,
  "namespace::autoclean" => 0,
  "parent" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



